code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/filter/genericobject/EnvAdminGenericObjectFilter.java

```diff
@@ -0,0 +1,123 @@
+        Object existing = criteria.get(key);
+
+        if (existing instanceof List) {
+            List<Object> newCondition = new ArrayList<Object>();
+            newCondition.add(value);
+            newCondition.addAll((List<?>)existing);
+            criteria.put(key, newCondition);
+        } else if (existing == null) {
+            criteria.put(key, value);
+        } else {
+            criteria.put(key, Arrays.asList(value, existing));
+        }
+        return criteria;
+    }
+
+    private boolean isEnvAdminGenericObject(String type, String id, ApiRequest request, ResourceManager next) {
+        Object obj = next.getById(type, id, new ListOptions(request));
+        if(obj != null && obj instanceof GenericObjectRecord){
+            if(GenericObjectConstants.ENV_ADMIN_TYPE.equals(((GenericObjectRecord)obj).getKind())){
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public Object getById(String type, String id, ListOptions options, ResourceManager next) {
+        Object obj = next.getById(type, id, options);
+        if(obj != null && obj instanceof GenericObjectRecord){
+            if(GenericObjectConstants.ENV_ADMIN_TYPE.equals(((GenericObjectRecord)obj).getKind())){
+                return null;
+            }
+        }
+        return obj;
+    }
+
+    @Override
+    public Object getLink(String type, String id, String link, ApiRequest request, ResourceManager next) {
+        if(isEnvAdminGenericObject(type, id, request, next)){
+            return null;
+        }
+        return next.getLink(type, id, link, request);
+    }
+
+    @Override
+    public Object list(String type, ApiRequest request, ResourceManager next) {
+        Map<Object, Object> criteria = setfilterEnvAdminCondition(new LinkedHashMap<Object, Object>(request.getConditions()));
+        return next.list(type, criteria, new ListOptions(request));
+    }
+
+    @Override
+    public List<?> list(String type, Map<Object, Object> criteria, ListOptions options, ResourceManager next) {
+        return next.list(type, setfilterEnvAdminCondition(criteria), options);
+    }
+
+    @Override
+    public Object create(String type, ApiRequest request, ResourceManager next) {
+        Map<Object, Object> properties = CollectionUtils.toMap(request.getRequestObject());
+        if(GenericObjectConstants.ENV_ADMIN_TYPE.equals(properties.get(ObjectMetaDataManager.KIND_FIELD))){
+            return null;
+        }
+        return next.create(type, request);
+    }
+
+    @Override
+    public Object update(String type, String id, ApiRequest request, ResourceManager next) {
+        if(isEnvAdminGenericObject(type, id, request, next)){
+            return null;
+        }
+        return next.update(type, id, request);
+    }
+
+    @Override
+    public Object delete(String type, String id, ApiRequest request, ResourceManager next) {
+        if(isEnvAdminGenericObject(type, id, request, next)){
```

Comment:
are you making it impossible to delete envAdminGenObj?

Reply:
This is a filter intercepting genericobject API. Deletion can be done via envAdminGenObj API where type=envAdmGO